### PR TITLE
Simd 118: add recalculation method and use when booting from snapshot

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
As per SIMD-0118, partitioned rewards should be recalculated on boot from snapshot. Partition and initial calculation info is available in the EpochRewards sysvar, and as of #1138, `Bank::get_epoch_reward_calculate_param_info` provides vote-account `credits_observed` from the EpochStakes cache. This is everything needed to recalculate undistributed stake-reward partitions.

#### Summary of Changes
Add `Bank::recalculate_partitioned_rewards` method. Recalculated partitions could differ from the original calculation in two ways:
1. Block partitions that have already been rewarded will be empty. This is because stake-account state is pulled from the StakesCache, which is modified as distribution occurs. (Specifically, stake `credits_observed` are updated to the new amount, so subsequent calculations determine no distribution needed.) Since only upcoming partitions need action (distribution), this is fine.
2. Rewards within a particular partition may appear in a different order. The runtime makes no guarantees about reward order, either in actual account updates or in block metadata storage, so this is fine.

Use recalculate_partitioned_rewards in `Bank::new_from_fields` to populate `Bank::epoch_reward_status`

### CODE REVIEW

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs [hunk 1]
```diff
@@ -512,6 +513,72 @@ impl Bank {
...
+        let epoch_rewards_sysvar = self.get_epoch_rewards_sysvar();
+        if epoch_rewards_sysvar.active {
+            let stake_rewards_by_partition = self.recalculate_stake_rewards(
+                &epoch_rewards_sysvar,
+                reward_calc_tracer,
+                thread_pool,
+            );
+            let start_block_height = epoch_rewards_sysvar
+                .distribution_starting_block_height
+                .saturating_sub(self.get_reward_calculation_num_blocks());
```
Comment (jstarry) [inline runtime/src/bank/partitioned_epoch_rewards/calculation.rs:535]: This isn't specific to this PR but I just want to call out that it's kind of awkward that this `start_block_height` is only used for distribution but it actually represents when _calculation_ was started. This differs from the `distribution_starting_block_height` in the epoch rewards sysvar despite the comment for this field stating that this represents the height "at which rewards distribution began":

```rust
pub(crate) struct StartBlockHeightAndRewards {
    /// the block height of the slot at which rewards distribution began
    pub(crate) start_block_height: u64,
```

Can you fix this up in a follow up please?

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs [hunk 2]
```diff
@@ -512,6 +513,72 @@ impl Bank {
 
         (points > 0).then_some(PointValue { rewards, points })
     }
+
+    /// If rewards are active, recalculates partitioned stake rewards and stores
+    /// a new Bank::epoch_reward_status. This method assumes that vote rewards
+    /// have already been calculated and delivered, and *only* recalculates
+    /// stake rewards
```
Comment (jstarry) [inline runtime/src/bank/partitioned_epoch_rewards/calculation.rs:520]: So right now `reward_calculation_num_blocks` is always `1` but in theory it could be adjusted and it would break this method's assumptions because if we started up from a snapshot in the middle of rewards calculation, the epoch rewards sysvar would be "active" even though rewards calculation hadn't finished yet. What do you think about removing that config var and replacing it with a constant in a follow-up?

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs [hunk 3]
```diff
@@ -512,6 +513,72 @@ impl Bank {
...
+    pub(in crate::bank) fn recalculate_partitioned_rewards(
+        &mut self,
+        reward_calc_tracer: Option<impl RewardCalcTracer>,
```
Comment (jstarry) [inline runtime/src/bank/partitioned_epoch_rewards/calculation.rs:523]: nit: how about using `null_tracer` when calling `calculate_stake_vote_rewards` instead of passing it through all the functions?

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs [hunk 4]
```diff
@@ -512,6 +513,72 @@ impl Bank {
...
+            StakeRewardCalculation {
+                mut stake_rewards, ..
+            },
+        ) = self.calculate_stake_vote_rewards(
+            &reward_calculate_param,
+            rewarded_epoch,
+            point_value,
+            thread_pool,
+            reward_calc_tracer,
+            &mut RewardsMetrics::default(), // This is required, but not reporting anything at the moment
```
Comment (jstarry) [inline runtime/src/bank/partitioned_epoch_rewards/calculation.rs:573]: Seems useful to report these metrics, are you planning to add this later?

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs [hunk 5]
```diff
@@ -512,6 +513,72 @@ impl Bank {
...
+        );
+        drop(stakes);
+        hash_rewards_into_partitions(
+            std::mem::take(&mut stake_rewards),
```
Comment (jstarry) [inline runtime/src/bank/partitioned_epoch_rewards/calculation.rs]: nit: `stake_rewards` can be passed by value here

File: runtime/src/bank.rs [hunk 6]
```diff
@@ -1652,6 +1652,15 @@ impl Bank {
             HashSet::default(),
         );
 
+        let (thread_pool, _thread_pool_time) = measure!(
```
Comment (jstarry) [inline runtime/src/bank.rs]: nit: if we're not reporting the timing, let's drop the `measure!` usage

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs [hunk 7]
```diff
@@ -541,13 +617,25 @@ mod tests {
...
+    ) -> (Bank, Vec<Pubkey>, Vec<Pubkey>) {
+        create_reward_bank(
+            expected_num_delegations,
+            PartitionedEpochRewardsConfig::default().stake_account_stores_per_block,
+        )
+    }
+
+    fn create_reward_bank(
+        expected_num_delegations: usize,
+        rewards_per_block: u64,
```
Comment (jstarry) [inline runtime/src/bank/partitioned_epoch_rewards/calculation.rs]: nit: this name is a little confusing, I prefer `stake_account_stores_per_block`

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs [hunk 8]
```diff
@@ -861,4 +969,229 @@ mod tests {
...
+        let rewarded_epoch = bank.epoch();
+
+        // Advance to next epoch boundary to update EpochStakes
+        let (bank, bank_forks) = bank.wrap_with_bank_forks_for_tests();
+        let bank = new_bank_from_parent_with_bank_forks(
+            &bank_forks,
+            bank,
+            &Pubkey::default(),
+            SLOTS_PER_EPOCH,
+        );
```
Comment (jstarry) [inline runtime/src/bank/partitioned_epoch_rewards/calculation.rs]: nit: it would be nice to put all of this in `create_reward_bank` since it looks like it's always advanced to the next epoch boundary

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs [hunk 9]
```diff
@@ -861,4 +969,229 @@ mod tests {
...
+            bank.recalculate_stake_rewards(&epoch_rewards_sysvar, null_tracer(), &thread_pool);
+        assert_eq!(expected_stake_rewards.len(), recalculated_rewards.len());
+        compare_stake_rewards(&expected_stake_rewards, &recalculated_rewards);
+
+        // Advance to first distribution slot
+        let bank = new_bank_from_parent_with_bank_forks(
+            &bank_forks,
+            bank,
+            &Pubkey::default(),
+            SLOTS_PER_EPOCH + 1,
```
Comment (jstarry) [inline runtime/src/bank/partitioned_epoch_rewards/calculation.rs]: nit: `bank.slot() + 1`

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs [hunk 10]
```diff
@@ -861,4 +969,229 @@ mod tests {
...
+        else {
+            panic!("{:?} not active", bank.epoch_reward_status);
+        };
+        assert_eq!(expected_starting_block_height, start_block_height);
+        assert_eq!(expected_stake_rewards.len(), recalculated_rewards.len());
+        compare_stake_rewards(&expected_stake_rewards, recalculated_rewards);
+
+        // Advance to first distribution slot
+        let mut bank =
+            Bank::new_from_parent(Arc::new(bank), &Pubkey::default(), SLOTS_PER_EPOCH + 1);
```
Comment (jstarry) [inline runtime/src/bank/partitioned_epoch_rewards/calculation.rs:1142]: optional but maybe we increment by two slots to ensure that block height is used to index distribution rather than slot number

File: runtime/src/bank.rs [hunk 11]
```diff
@@ -1636,7 +1636,7 @@ impl Bank {
             accounts_data_size_initial,
             accounts_data_size_delta_on_chain: AtomicI64::new(0),
             accounts_data_size_delta_off_chain: AtomicI64::new(0),
-            epoch_reward_status: fields.epoch_reward_status,
+            epoch_reward_status: EpochRewardStatus::default(),
```
Comment (jstarry) [inline runtime/src/bank.rs:1640]: Can this be changed in a separate PR? It doesn't seem related to the rest of your changes.. (snapshot tests too)

Comment (CriesofCarrots) [reply runtime/src/bank.rs:1640]: The commit that does this can be moved to a separate PR. Unfortunately, the snapshot tests do have to be handled in some fashion. They don't initialize the feature set fully or populate the EpochRewards sysvar, so the new `epoch_reward_status` always reports inactive and fails the various comparison asserts after `Bank::new_from_fields` runs.

I hacked at it a little to see if it would be simple to fix, but it wasn't really. I didn't think it was worth much time, since it no longer matters if the snapshot returns the right epoch-rewards data after this PR.

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs (hunk 2 above)
Comment (CriesofCarrots) [reply runtime/src/bank/partitioned_epoch_rewards/calculation.rs:520]: Yeah, makes a lot of sense. I'll also do a scan to see how broken things will be if calculation num blocks increase; maybe there are some small things we can do to reduce that brittleness.

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs (hunk 1 above)
Comment (CriesofCarrots) [reply runtime/src/bank/partitioned_epoch_rewards/calculation.rs:535]: Yes, totally. I should have fixed it already 😞 ; it is driving me bonkers in tests.

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs (hunk 4 above)
Comment (CriesofCarrots) [reply runtime/src/bank/partitioned_epoch_rewards/calculation.rs:573]: Yep. I want to take a look at the metrics separately with the whole implementation in mind. The reporting cadence will need to change, and I don't think all the fields will still be relevant.

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs (hunk 3 above)
Comment (CriesofCarrots) [reply runtime/src/bank/partitioned_epoch_rewards/calculation.rs:523]: I think a ledger-tool user could want tracer populated in `calculate_stake_vote_rewards`, no?
It's true that currently it isn't ever populated for `Bank::new_from_fields()` (and hence for this method). I could remove it here, but I thought we might want it upstream for testing.

**Comment by CriesofCarrots** [comment]
Rebased to pick up tip of master; changes all in the new, last 4 commits

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs (hunk 2 above)
Comment (jstarry) [reply runtime/src/bank/partitioned_epoch_rewards/calculation.rs:520]: Cool, just to be clear I think we can commit to keeping calculation in the first block for now.. SIMD-0118 doesn't mention that the number of calculation blocks is configurable at all. So I'm strongly in favor of ripping out anything that makes it configurable and lean into assuming that it's a single block for now (and therefore if the epoch sysvar is active in a frozen bank, we are definitely in distribution). If we ever decide to extend it over multiple blocks, it will require a new SIMD and careful implementation beyond just updating a constant.

**Review by jstarry — DISMISSED** [review]

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs [hunk 12]
```diff
@@ -512,6 +513,67 @@ impl Bank {
...
+            );
+            let start_block_height = epoch_rewards_sysvar
+                .distribution_starting_block_height
+                .saturating_sub(self.get_reward_calculation_num_blocks());
+            self.set_epoch_reward_status_active(start_block_height, stake_rewards_by_partition);
+        }
+    }
+
+    /// Returns a vector of partitioned stake rewards. StakeRewards are
+    /// recalculated active EpochRewards sysvar, vote accounts from EpochStakes,
```
Comment (joncinque) [inline runtime/src/bank/partitioned_epoch_rewards/calculation.rs]: micro-nit
```suggestion
    /// recalculated from active EpochRewards sysvar, vote accounts from EpochStakes,
```

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs [hunk 13]
```diff
@@ -512,6 +513,67 @@ impl Bank {
...
+
+        let point_value = PointValue {
+            rewards: epoch_rewards_sysvar.total_rewards,
+            points: epoch_rewards_sysvar.total_points,
+        };
+
+        let stakes = self.stakes_cache.stakes();
+        let reward_calculate_param = self.get_epoch_reward_calculate_param_info(&stakes);
+
+        let (_, StakeRewardCalculation { stake_rewards, .. }) = self.calculate_stake_vote_rewards(
```
Comment (joncinque) [inline runtime/src/bank/partitioned_epoch_rewards/calculation.rs:567]: The comment in `recalculate_partitioned_rewards` makes it clear, but could you add a comment here saying that the other outputs from this will *not* be correct? Ie. the voting rewards and total rewards won't contain already distributed rewards.

Just to stop someone in case they think to use those parameters later.

**Review by joncinque — DISMISSED** [review]
Looks great! Just a couple of tiny things, can be addressed separately if more convenient

**Comment by CriesofCarrots** [comment]
> Looks great! Just a couple of tiny things, can be addressed separately if more convenient

Thanks! Went ahead and pushed those here as the last commit. Rebasing to ensure nothing weird upstream.

**Review by joncinque — APPROVED** [review]

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs (hunk 1 above)
Comment (CriesofCarrots) [reply runtime/src/bank/partitioned_epoch_rewards/calculation.rs:535]: To be thorough, done here: https://github.com/anza-xyz/agave/pull/1239

File: runtime/src/bank/partitioned_epoch_rewards/calculation.rs (hunk 2 above)
Comment (CriesofCarrots) [reply runtime/src/bank/partitioned_epoch_rewards/calculation.rs:520]: Follow-up here: https://github.com/anza-xyz/agave/pull/1251
